LocalLang/SemanticsLemmas.lean

```diff
@@ -8,14 +8,71 @@ instance :
       (Relation.ReflTransGen.single st₁) (Relation.ReflTransGen.single st₂)
 
 @[simp] lemma addBindings_single {e xe : Expr}
-    : e.addBindings [(x, xe)] = Expr.letIn x xe e := by sorry
+    : e.addBindings [(x, xe)] = Expr.letIn x xe e := by
+  simp [Expr.addBindings]
 
-lemma SmallStep.hole_step : HeadSmallStep defs V e₁ e₂ → SmallStep defs V e₁ e₂ := by sorry
+lemma SmallStep.hole_step : HeadSmallStep defs V e₁ e₂ → SmallStep defs V e₁ e₂ := by
+  intro hstep
+  exact .ctx_step .hole (by rw [Ctx.fill]) (by rw [Ctx.fill]) hstep
+
+lemma nestCtx_preserves_updateEnv (ctx' ctx : Ctx) {V : Env}
+  : ctx'.updateEnv (ctx.updateEnv V) = (ctx.fillWithCtx ctx').updateEnv V := by
+  induction ctx generalizing V <;> simp only [Ctx.updateEnv, Ctx.fillWithCtx] <;> (
+    expose_names
+    apply a_ih
+  )
+
+lemma nestCtx_preserves_fill (ctx' ctx : Ctx) {e : Expr}
+  : ctx.fill (ctx'.fill e) = (ctx.fillWithCtx ctx').fill e := by
+  induction ctx <;> simp only [Ctx.fill, Ctx.fillWithCtx] <;> congr
 
 lemma SmallStep.with_ctx (ctx : Ctx)
     : (e₁' = ctx.fill e₁) → (e₂' = ctx.fill e₂)
-    → SmallStep defs (ctx.updateEnv V) e₁ e₂ → SmallStep defs V e₁' e₂' := by sorry
+    → SmallStep defs (ctx.updateEnv V) e₁ e₂ → SmallStep defs V e₁' e₂' := by
+  intro e₁'_eq e₂'_eq step
+  let ⟨ctx', e₁_eq, e₂_eq, hstep⟩ := step
+  rename_i e₁_h e₂_h
+  rw [nestCtx_preserves_updateEnv ctx' ctx] at hstep
+  apply ctx_step (ctx.fillWithCtx ctx') ?_ ?_ hstep
+  · rw [e₁'_eq, e₁_eq]; apply nestCtx_preserves_fill
+  · rw [e₂'_eq, e₂_eq]; apply nestCtx_preserves_fill
 
 lemma SmallSteps.with_ctx (ctx : Ctx)
     : (e₁' = ctx.fill e₁) → (e₂' = ctx.fill e₂)
-    → SmallSteps defs (ctx.updateEnv V) e₁ e₂ → SmallSteps defs V e₁' e₂' := by sorry
+    → SmallSteps defs (ctx.updateEnv V) e₁ e₂ → SmallSteps defs V e₁' e₂' := by
+  intro e₁'_eq e₂'_eq steps
+  rw [e₁'_eq, e₂'_eq]
+  apply Relation.ReflTransGen.lift (ctx.fill) (by
+    intro e e' step
+    exact SmallStep.with_ctx ctx rfl rfl step
+  ) steps
```

Comment:
Please avoid this kind of thing, it's very hard to read.  You can use `?_` instead.